src/zkp/pifac.rs

```diff
@@ -156,64 +201,78 @@ impl Proof for PiFacProof {
+            let rhs = input.setup_params.scheme().combine(
+                &self.p_mask_commitment,
+                &self.p_commitment,
+                &e,
+            );
             lhs == rhs
         };
-        if !eq_check_1 {
-            error!("eq_check_1 failed");
+        if !masked_p_commitment_is_valid {
+            error!("masked_p_commitment_is_valid failed");
             return Err(InternalError::ProtocolError);
         }
 
-        let eq_check_2 = {
-            let lhs = input.setup_params.scheme().reconstruct(&self.z2, &self.w2);
-            let rhs = input.setup_params.scheme().combine(&self.B, &self.Q, &e);
+        let masked_q_commitment_is_valid = {
+            let lhs = input
+                .setup_params
+                .scheme()
+                .reconstruct(&self.q_masked, &self.masked_q_commitment_randomness);
+            let rhs = input.setup_params.scheme().combine(
+                &self.q_mask_commitment,
+                &self.q_commitment,
+                &e,
+            );
             lhs == rhs
         };
-        if !eq_check_2 {
-            error!("eq_check_2 failed");
+        if !masked_q_commitment_is_valid {
+            error!("masked_q_commitment_is_valid failed");
             return Err(InternalError::ProtocolError);
         }
 
-        let eq_check_3 = {
-            let R = input
-                .setup_params
-                .scheme()
-                .reconstruct(&input.N0, self.sigma.as_masked());
-            let lhs = input
+        let modulus_links_provided_factors = {
+            let reconstructed_commitment = input
                 .setup_params
                 .scheme()
-                .reconstruct_with_commitment(&self.Q, &self.z1, &self.v);
-            let rhs = input.setup_params.scheme().combine(&self.T, &R, &e);
+                .reconstruct(&input.modulus, self.link_randomness.as_masked());
+            let lhs = input.setup_params.scheme().reconstruct_with_commitment(
+                &self.q_commitment,
+                &self.p_masked,
+                &self.masked_p_link,
+            );
+            let rhs = input.setup_params.scheme().combine(
+                &self.q_link_commitment,
+                &reconstructed_commitment,
+                &e,
+            );
             lhs == rhs
         };
-        if !eq_check_3 {
-            error!("eq_check_3 failed");
+        if !modulus_links_provided_factors {
+            error!("modulus_links_provided_factors failed");
             return Err(InternalError::ProtocolError);
         }
 
-        let sqrt_N0 = sqrt(&input.N0);
-        // 2^{ELL + EPSILON}
-        let two_ell_eps = BigNumber::one() << (ELL + EPSILON);
-        // 2^{ELL + EPSILON} * sqrt(N_0)
-        let z_bound = &sqrt_N0 * &two_ell_eps;
-        if self.z1 < -z_bound.clone() || self.z1 > z_bound {
+        if crate::utils::within_bound_by_size(&self.p_masked, ELL + EPSILON) {
             error!("self.z1 > z_bound check failed");
```

Comment:
todo: Make these logs more useful. If this fails it means that `p` is out of range. Ditto below for `z2`/`q`.